File: backend/src/routes/coins.rs
use axum::{
    extract::State,
    response::Json,
    routing::{get, post},
    Router,
};
use chrono::Utc;
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use validator::Validate;

use crate::{
    error::{AppError, Result},
    models::Claims,
    AppState,
};

#[derive(Debug, Deserialize, Validate)]
pub struct CreateTransactionRequest {
    #[validate(range(min = 1))]
    pub amount: i32,
    #[validate(length(min = 1, max = 50))]
    pub transaction_type: String,
    #[validate(length(min = 1, max = 200))]
    pub description: String,
}

#[derive(Debug, Serialize)]
pub struct CoinBalance {
    pub balance: i64,
    pub total_earned: i64,
    pub total_spent: i64,
    pub last_updated: chrono::DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct CoinTransaction {
    pub id: Uuid,
    pub amount: i32,
    pub transaction_type: String,
    pub description: String,
    pub transaction_date: chrono::DateTime<Utc>,
}

pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/balance", get(get_balance))
        .route("/transactions", get(get_coin_transactions))
        .route("/transaction", post(create_transaction))
}

/// Get coin balance for the couple
/// GET /api/coins/balance
async fn get_balance(
    State(state): State<AppState>,
    claims: Claims,
) -> Result<Json<CoinBalance>> {
    tracing::debug!("Fetching coin balance for user {}", claims.sub);
    
    let user_id = Uuid::parse_str(&claims.sub)
        .map_err(|_| AppError::Auth("無效的用戶ID".to_string()))?;

    // Find the user's couple
    let couple = sqlx::query!(
        "SELECT id FROM couples WHERE user1_id = $1 OR user2_id = $1",
        user_id
    )
    .fetch_optional(&state.db.pool)
    .await?;

    if couple.is_none() {
        tracing::warn!("User {} attempted to fetch coin balance but has no couple", user_id);
        return Err(AppError::NotFound("您還沒有配對。請先創建情侶檔案。".to_string()));
    }

    let couple_id = couple.unwrap().id;
    tracing::debug!("Found couple {} for user {}", couple_id, user_id);

    // Get balance from the view
    let balance_result = sqlx::query!(
        "SELECT balance, last_updated FROM coin_balances WHERE couple_id = $1",
        couple_id
    )
    .fetch_optional(&state.db.pool)
    .await?;

    tracing::debug!("Raw balance result for couple {}: {:?}", couple_id, balance_result);

    let balance = balance_result.as_ref().map(|r| r.balance).unwrap_or(Some(0)).unwrap_or(0);
    let last_updated = balance_result.as_ref().map(|r| r.last_updated).unwrap_or_else(|| Some(Utc::now())).unwrap_or_else(|| Utc::now());

    // Get total earned and spent
    let totals = sqlx::query!(
        "SELECT 
            COALESCE(SUM(CASE WHEN transaction_type = 'earn' THEN amount ELSE 0 END), 0) as total_earned,
            COALESCE(SUM(CASE WHEN transaction_type = 'spend' THEN amount ELSE 0 END), 0) as total_spent
         FROM coin_transactions 
         WHERE couple_id = $1",
        couple_id
    )
    .fetch_one(&state.db.pool)
    .await?;

    tracing::debug!(
        "Coin stats for couple {}: balance={}, earned={}, spent={}", 
        couple_id, 
        balance,
        totals.total_earned.unwrap_or(0),
        totals.total_spent.unwrap_or(0)
    );

    Ok(Json(CoinBalance {
        balance,
        total_earned: totals.total_earned.unwrap_or(0),
        total_spent: totals.total_spent.unwrap_or(0),
        last_updated,
    }))
}

/// Get coin transaction history
/// GET /api/coins/transactions
async fn get_coin_transactions(
    State(state): State<AppState>,
    claims: Claims,
) -> Result<Json<Vec<CoinTransaction>>> {
    tracing::debug!("Fetching coin transactions for user {}", claims.sub);
    
    let user_id = Uuid::parse_str(&claims.sub)
        .map_err(|_| AppError::Auth("無效的用戶ID".to_string()))?;

    // Find the user's couple
    let couple = sqlx::query!(
        "SELECT id FROM couples WHERE user1_id = $1 OR user2_id = $1",
        user_id
    )
    .fetch_optional(&state.db.pool)
    .await?;

    if couple.is_none() {
        tracing::warn!("User {} attempted to fetch coin transactions but has no couple", user_id);
        return Err(AppError::NotFound("您還沒有配對。請先創建情侶檔案。".to_string()));
    }

    let couple_id = couple.unwrap().id;
    tracing::debug!("Found couple {} for user {}", couple_id, user_id);

    // Get transaction history
    let transactions = sqlx::query!(
        "SELECT id, amount, transaction_type, description, transaction_date
         FROM coin_transactions 
         WHERE couple_id = $1
         ORDER BY transaction_date DESC
         LIMIT 50",
        couple_id
    )
    .fetch_all(&state.db.pool)
    .await?;

    tracing::debug!("Found {} transactions for couple {}", transactions.len(), couple_id);

    let response_transactions: Vec<CoinTransaction> = transactions.into_iter().map(|tx| {
        CoinTransaction {
            id: tx.id,
            amount: tx.amount,
            transaction_type: tx.transaction_type,
            description: tx.description,
            transaction_date: tx.transaction_date.unwrap_or_else(|| Utc::now()),
        }
    }).collect();

    Ok(Json(response_transactions))
}

/// Record a coin transaction
/// POST /api/coins/transaction
async fn create_transaction(
    State(state): State<AppState>,
    claims: Claims,
    Json(payload): Json<CreateTransactionRequest>,
) -> Result<Json<CoinTransaction>> {
    // Validate input
    payload.validate().map_err(|e| {
        AppError::Validation(format!("驗證失敗: {}", e))
    })?;

    let user_id = Uuid::parse_str(&claims.sub)
        .map_err(|_| AppError::Auth("無效的用戶ID".to_string()))?;

    // Find the user's couple
    let couple = sqlx::query!(
        "SELECT id FROM couples WHERE user1_id = $1 OR user2_id = $1",
        user_id
    )
    .fetch_optional(&state.db.pool)
    .await?
    .ok_or_else(|| AppError::NotFound("您還沒有配對。請先創建情侶檔案。".to_string()))?;

    // Validate transaction type
    if payload.transaction_type != "earn" && payload.transaction_type != "spend" {
        return Err(AppError::Validation("交易類型必須是 'earn' 或 'spend'".to_string()));
    }

    // For spend transactions, check if user has enough balance
    if payload.transaction_type == "spend" {
        let current_balance = sqlx::query!(
            "SELECT COALESCE(balance, 0) as balance FROM coin_balances WHERE couple_id = $1",
            couple.id
        )
        .fetch_optional(&state.db.pool)
        .await?
        .map(|r| r.balance)
        .unwrap_or(Some(0)).unwrap_or(0);

        if current_balance < payload.amount as i64 {
            return Err(AppError::Validation("餘額不足".to_string()));
        }
    }

    // Create transaction
    let transaction_id = Uuid::new_v4();
    let now = Utc::now();

    sqlx::query!(
        "INSERT INTO coin_transactions (id, couple_id, amount, transaction_type, description, transaction_date)
         VALUES ($1, $2, $3, $4, $5, $6)",
        transaction_id,
        couple.id,
        payload.amount,
        payload.transaction_type,
        payload.description,
        now
    )
    .execute(&state.db.pool)
    .await?;

    Ok(Json(CoinTransaction {
        id: transaction_id,
        amount: payload.amount,
        transaction_type: payload.transaction_type,
        description: payload.description,
        transaction_date: now,
    }))
} 

File: routes/coins.js
const express = require('express');
const { body, validationResult } = require('express-validator');
const { v4: uuidv4 } = require('uuid');
const db = require('../database/db');
const { authenticateToken } = require('../middleware/auth');
const { logInfo, logWarn, logError } = require('../lib/logger');

const router = express.Router();

// All coin routes require authentication
router.use(authenticateToken);

// Get coin balance for couple
router.get('/balance', async (req, res) => {
  try {
    const userId = req.user.id;

    // Find user's couple
    const coupleResult = await db.query(
      'SELECT id FROM couples WHERE user1_id = $1 OR user2_id = $1',
      [userId]
    );

    if (coupleResult.rows.length === 0) {
      return res.status(404).json({
        success: false,
        message: '您還沒有情侶關係'
      });
    }

    const coupleId = coupleResult.rows[0].id;

    // Calculate balance from transactions
    const result = await db.query(`
      SELECT
        COALESCE(SUM(CASE WHEN transaction_type = 'earn' THEN amount ELSE -amount END), 0) as balance
      FROM coin_transactions
      WHERE couple_id = $1
    `, [coupleId]);

    const balance = parseInt(result.rows[0].balance) || 0;

    res.json({
      success: true,
      balance
    });

  } catch (error) {
    logError('Get coin balance failed', { err: error.message, stack: error.stack });
    res.status(500).json({
      success: false,
      message: '獲取金幣餘額失敗'
    });
  }
});

// Get coin transaction history for couple
router.get('/transactions', async (req, res) => {
  try {
    const userId = req.user.id;
    const { page = 1, limit = 20 } = req.query;

    const offset = (page - 1) * limit;

    // Find user's couple
    const coupleResult = await db.query(
      'SELECT id FROM couples WHERE user1_id = $1 OR user2_id = $1',
      [userId]
    );

    if (coupleResult.rows.length === 0) {
      return res.status(404).json({
        success: false,
        message: '您還沒有情侶關係'
      });
    }

    const coupleId = coupleResult.rows[0].id;

    // Get total count
    const countResult = await db.query(
      'SELECT COUNT(*) as total FROM coin_transactions WHERE couple_id = $1',
      [coupleId]
    );
    const total = parseInt(countResult.rows[0].total);

    // Get transactions
    const result = await db.query(`
      SELECT id, amount, transaction_type, earned_from, spent_on, description, transaction_date
      FROM coin_transactions
      WHERE couple_id = $1
      ORDER BY transaction_date DESC
      LIMIT $2 OFFSET $3
    `, [coupleId, parseInt(limit), offset]);

    res.json({
      success: true,
      transactions: result.rows,
      total,
      page: parseInt(page),
      limit: parseInt(limit)
    });

  } catch (error) {
    logError('Get coin transactions failed', { err: error.message, stack: error.stack });
    res.status(500).json({
      success: false,
      message: '獲取金幣交易記錄失敗'
    });
  }
});

// Add coins (for achievements, daily rewards, etc.)
router.post('/add', [
  body('amount')
    .isInt({ min: 1 })
    .withMessage('金幣數量必須是正整數'),
  body('reason')
    .isLength({ min: 1, max: 200 })
    .withMessage('原因描述必須在1-200個字符之間'),
  body('earned_from')
    .optional()
    .isLength({ max: 100 })
    .withMessage('獲得來源不能超過100個字符')
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        success: false,
        message: '驗證失敗',
        errors: errors.array()
      });
    }

    const userId = req.user.id;
    const { amount, reason, earned_from } = req.body;

    // Find user's couple
    const coupleResult = await db.query(
      'SELECT id FROM couples WHERE user1_id = $1 OR user2_id = $1',
      [userId]
    );

    if (coupleResult.rows.length === 0) {
      return res.status(404).json({
        success: false,
        message: '您還沒有情侶關係'
      });
    }

    const coupleId = coupleResult.rows[0].id;

    // Record transaction
    const transactionId = uuidv4();
    await db.query(`
      INSERT INTO coin_transactions (id, couple_id, amount, transaction_type, earned_from, description, transaction_date)
      VALUES ($1, $2, $3, 'earn', $4, $5, NOW())
    `, [transactionId, coupleId, amount, earned_from, reason]);

    // Get updated balance
    const balanceResult = await db.query(`
      SELECT COALESCE(SUM(CASE WHEN transaction_type = 'earn' THEN amount ELSE -amount END), 0) as balance
      FROM coin_transactions
      WHERE couple_id = $1
    `, [coupleId]);
    const newBalance = parseInt(balanceResult.rows[0].balance);

    logInfo('Added coins', { coupleId, amount, reason });

    res.json({
      success: true,
      message: `獲得 ${amount} 金幣！`,
      amount,
      new_balance: newBalance
    });

  } catch (error) {
    logError('Add coins failed', { err: error.message, stack: error.stack });
    res.status(500).json({
      success: false,
      message: '添加金幣失敗'
    });
  }
});

// Spend coins
router.post('/spend', [
  body('amount')
    .isInt({ min: 1 })
    .withMessage('金幣數量必須是正整數'),
  body('reason')
    .isLength({ min: 1, max: 200 })
    .withMessage('原因描述必須在1-200個字符之間'),
  body('transaction_type')
    .isIn(['purchase', 'unlock', 'gift', 'other'])
    .withMessage('交易類型無效')
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        success: false,
        message: '驗證失敗',
        errors: errors.array()
      });
    }

    const userId = req.user.id;
    const { amount, reason, transaction_type } = req.body;

    const coupleResult = await db.query(
      'SELECT id FROM couples WHERE user1_id = $1 OR user2_id = $1',
      [userId]
    );

    if (coupleResult.rows.length === 0) {
      return res.status(404).json({
        success: false,
        message: '您還沒有情侶關係'
      });
    }

    const coupleId = coupleResult.rows[0].id;

    const balanceResult = await db.query(`
      SELECT COALESCE(SUM(CASE WHEN transaction_type = 'earn' THEN amount ELSE -amount END), 0) as balance
      FROM coin_transactions
      WHERE couple_id = $1
    `, [coupleId]);
    const currentBalance = parseInt(balanceResult.rows[0].balance) || 0;

    if (currentBalance < amount) {
      return res.status(400).json({
        success: false,
        message: '金幣餘額不足',
        current_balance: currentBalance,
        required: amount
      });
    }

    const transactionId = uuidv4();
    await db.query(`
      INSERT INTO coin_transactions (id, couple_id, amount, transaction_type, spent_on, description, transaction_date)
      VALUES ($1, $2, $3, 'spend', $4, $5, NOW())
    `, [transactionId, coupleId, amount, transaction_type, reason]);

    const updatedBalanceResult = await db.query(`
      SELECT COALESCE(SUM(CASE WHEN transaction_type = 'earn' THEN amount ELSE -amount END), 0) as balance
      FROM coin_transactions
      WHERE couple_id = $1
    `, [coupleId]);
    const newBalance = parseInt(updatedBalanceResult.rows[0].balance);

    logInfo('Coins spent', { coupleId, amount, reason });

    res.json({
      success: true,
      message: `花費 ${amount} 金幣`,
      amount,
      new_balance: newBalance
    });

  } catch (error) {
    logError('Spend coins failed', { err: error.message, stack: error.stack });
    res.status(500).json({
      success: false,
      message: '花費金幣失敗'
    });
  }
});

// Daily coin claim
router.post('/daily-claim', async (req, res) => {
  try {
    const userId = req.user.id;

    // Find user's couple
    const coupleResult = await db.query(
      'SELECT id FROM couples WHERE user1_id = $1 OR user2_id = $1',
      [userId]
    );

    if (coupleResult.rows.length === 0) {
      return res.status(404).json({
        success: false,
        message: '您還沒有情侶關係'
      });
    }

    const coupleId = coupleResult.rows[0].id;

    // Check if couple already claimed today
    const today = new Date().toISOString().split('T')[0];
    const claimResult = await db.query(`
      SELECT id FROM coin_transactions
      WHERE couple_id = $1
      AND earned_from = 'daily_reward'
      AND DATE(transaction_date) = $2
    `, [coupleId, today]);

    if (claimResult.rows.length > 0) {
      return res.status(409).json({
        success: false,
        message: '今日已經領取過每日金幣'
      });
    }

    const dailyAmount = 10; // 10 coins per day

    // Record daily reward transaction
    const transactionId = uuidv4();
    await db.query(`
      INSERT INTO coin_transactions (id, couple_id, amount, transaction_type, earned_from, description, transaction_date)
      VALUES ($1, $2, $3, 'earn', 'daily_reward', $4, NOW())
    `, [transactionId, coupleId, dailyAmount, '每日登入獎勵']);

    // Get updated balance
    const balanceResult = await db.query(`
      SELECT COALESCE(SUM(CASE WHEN transaction_type = 'earn' THEN amount ELSE -amount END), 0) as balance
      FROM coin_transactions
      WHERE couple_id = $1
    `, [coupleId]);
    const newBalance = parseInt(balanceResult.rows[0].balance);

    logInfo('Daily coins claimed', { coupleId, amount: dailyAmount });

    res.json({
      success: true,
      message: `領取每日 ${dailyAmount} 金幣！`,
      amount: dailyAmount,
      new_balance: newBalance
    });

  } catch (error) {
    logError('Daily coin claim failed', { err: error.message, stack: error.stack });
    res.status(500).json({
      success: false,
      message: '領取每日金幣失敗'
    });
  }
});

// Generic transaction endpoint (for frontend compatibility)
router.post('/transaction', [
  body('amount')
    .isInt({ min: 1 })
    .withMessage('金幣數量必須是正整數'),
  body('transaction_type')
    .isIn(['earn', 'spend'])
    .withMessage('交易類型必須是 earn 或 spend'),
  body('description')
    .isLength({ min: 1, max: 200 })
    .withMessage('描述必須在1-200個字符之間'),
  // Optional client-generated idempotency key (a UUID per user intent). When
  // present, a retried/duplicated request with the same key grants only once.
  body('idempotency_key')
    .optional({ nullable: true })
    .isString()
    .isLength({ min: 1, max: 128 })
    .withMessage('idempotency_key 格式錯誤')
], async (req, res) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        success: false,
        message: '驗證失敗',
        errors: errors.array()
      });
    }

    const userId = req.user.id;
    const { amount, transaction_type, description, idempotency_key } = req.body;

    // Find user's couple
    const coupleResult = await db.query(
      'SELECT id FROM couples WHERE user1_id = $1 OR user2_id = $1',
      [userId]
    );

    if (coupleResult.rows.length === 0) {
      return res.status(404).json({
        success: false,
        message: '您還沒有情侶關係'
      });
    }

    const coupleId = coupleResult.rows[0].id;

    // If spending, check balance first
    if (transaction_type === 'spend') {
      const balanceResult = await db.query(`
        SELECT COALESCE(SUM(CASE WHEN transaction_type = 'earn' THEN amount ELSE -amount END), 0) as balance
        FROM coin_transactions
        WHERE couple_id = $1
      `, [coupleId]);

      const currentBalance = parseInt(balanceResult.rows[0].balance) || 0;

      if (currentBalance < amount) {
        return res.status(400).json({
          success: false,
          message: '金幣餘額不足',
          current_balance: currentBalance,
          required: amount
        });
      }
    }

    // Record transaction. When an idempotency key is supplied, ON CONFLICT
    // makes a repeat of the same intent a no-op instead of a second grant —
    // the arbiter is the partial unique index from migration 085, so the WHERE
    // predicate must match it for inference. A null key never conflicts (each
    // keyless row inserts as before).
    const transactionId = uuidv4();
    const insertResult = await db.query(`
      INSERT INTO coin_transactions (id, couple_id, amount, transaction_type, description, transaction_date, earned_from, spent_on, idempotency_key)
      VALUES ($1, $2, $3, $4, $5, NOW(), $6, $7, $8)
      ON CONFLICT (idempotency_key) WHERE idempotency_key IS NOT NULL DO NOTHING
      RETURNING id
    `, [
      transactionId,
      coupleId,
      amount,
      transaction_type,
      description,
      transaction_type === 'earn' ? description : null,
      transaction_type === 'spend' ? description : null,
      idempotency_key || null
    ]);

    // A key was given but nothing inserted → this exact request already ran.
    // Return success idempotently (with the balance that already reflects it)
    // so a client retry is a harmless no-op, never a double grant or an error.
    const deduped = !!idempotency_key && insertResult.rows.length === 0;

    // Get updated balance
    const balanceResult = await db.query(`
      SELECT COALESCE(SUM(CASE WHEN transaction_type = 'earn' THEN amount ELSE -amount END), 0) as balance
      FROM coin_transactions
      WHERE couple_id = $1
    `, [coupleId]);
    const newBalance = parseInt(balanceResult.rows[0].balance);

    const action = transaction_type === 'earn' ? '獲得' : '花費';
    if (deduped) {
      logWarn('Coin transaction deduped by idempotency key', { coupleId, action, amount, idempotency_key });
    } else {
      logInfo('Coin transaction', { coupleId, action, amount, description });
    }

    res.json({
      success: true,
      message: `${action} ${amount} 金幣`,
      amount,
      transaction_type,
      new_balance: newBalance,
      deduped
    });

  } catch (error) {
    logError('Coin transaction failed', { err: error.message, stack: error.stack });
    res.status(500).json({
      success: false,
      message: '金幣交易失敗'
    });
  }
});

module.exports = router;
